Source/FakeItEasy/Core/EventHandlerArgumentProviderMap.cs

```diff
@@ -1,29 +1,24 @@
 ﻿namespace FakeItEasy.Core
 {
     using System;
+    using System.Collections.Concurrent;
     using System.Collections.Generic;
 
     internal class EventHandlerArgumentProviderMap
     {
-        private readonly Dictionary<Delegate, IEventRaiserArguments> map
-            = new Dictionary<Delegate, IEventRaiserArguments>(new EventRaiserDelegateComparer());
+        private readonly ConcurrentDictionary<Delegate, IEventRaiserArguments> map
+            = new ConcurrentDictionary<Delegate, IEventRaiserArguments>(new EventRaiserDelegateComparer());
 
         public void AddArgumentProvider(Delegate eventHandler, IEventRaiserArguments argumentProvider)
         {
-            lock (this.map)
-            {
-                this.map[eventHandler] = argumentProvider;
-            }
+            this.map[eventHandler] = argumentProvider;
         }
 
         public IEventRaiserArguments TakeArgumentProviderFor(Delegate eventHandler)
         {
-            lock (this.map)
-            {
-                var provider = this.map[eventHandler];
-                this.map.Remove(eventHandler);
-                return provider;
-            }
+            IEventRaiserArguments provider;
+            this.map.TryRemove(eventHandler, out provider);
+            return provider;

```

Comment:
I suspected that the public behaviour of package hasn't changed, thanks for confirming that.

So we know the way in which the type is currently used means the package works. The remaining question is, does this type make sense in isolation? I.e. does it represent a sensible encapsulation and do it's method signatures convey the right semantics? I appreciate it's an internal type, but these questions are still important, since *someone* still has to work with it (contributors).